ToolBox/PsychSignalDataConverter/PsychSignalDataDownloader.cs

```diff
@@ -0,0 +1,149 @@
+        
+        /// <summary>
+        /// Maximum amount of retries per data hour 
+        /// </summary>
+        public int MaxRetries = 5;
+        
+        /// <summary>
+        /// Downlods data from psychsignal
+        /// </summary>
+        /// <param name="rawDataDestination">Directory we write raw data to</param>
+        /// <param name="apiKey">PsychSignal API key</param>
+        /// <param name="dataSource">Data source (e.g. stocktwits,twitter_withretweets)</param>
+        public PsychSignalDataDownloader(string rawDataDestination, string apiKey, string dataSource)
+        {
+            _rawDataDestination = rawDataDestination;
+
+            _dataSource = dataSource;
+            _apiKey = apiKey;
+        }
+        
+        /// <summary>
+        /// Download the data from the given starting date to the ending date.
+        /// Note that if the ending date is in the same hour as the current time,
+        /// we will lower the <paramref name="endDateUtc" /> by one hour in order
+        /// to make sure that we only download complete, non-changing dataa
+        /// </summary>
+        /// <param name="startDateUtc">Starting date</param>
+        /// <param name="endDateUtc">Ending date</param>
+        public void Download(DateTime startDateUtc, DateTime endDateUtc)
+        {
+            if (startDateUtc < endDateUtc.AddDays(-15))
+            {
+                throw new ArgumentException("The starting date can only be at most 15 days from now");
+            }
+
+            Directory.CreateDirectory(_rawDataDestination);
+            var now = DateTime.UtcNow;
+
+            // Makes sure we only get final, non-changing data by checking if the current hour matches the 
+            // hour of the upper date bound and subtracting an hour from it if that is the case
+            if (new DateTime(now.Year, now.Month, now.Day, now.Hour, 0, 0) == new DateTime(endDateUtc.Year, endDateUtc.Month, endDateUtc.Day, endDateUtc.Hour, 0, 0))
+            {
+                endDateUtc = endDateUtc.AddHours(-1);
+            }
+            
+            // PsychSignal paginates data by hour
+            for (; startDateUtc < endDateUtc; startDateUtc = startDateUtc.AddHours(1))
+            {
+                var rawDataPath = Path.Combine(_rawDataDestination, $"{startDateUtc:yyyyMMdd_HH}_{_dataSource}.csv");
+                var rawDataPathTemp = Path.Combine(Path.GetTempPath(), $"{startDateUtc:yyyyMMdd_HH}_{_dataSource}.csv.tmp");
+                
+                // Don't download files we already have
+                if (File.Exists(rawDataPath))
+                {
+                    continue;
+                }
+                
+                // Retry in case a download failed
+                for (var retries = 0; retries < MaxRetries; retries++)
+                {
+                    // Psychsignal imposes very strict rate limits
+                    Thread.Sleep(10000);
```

Comment:
Consider using the `RateGate` type to handle rate limits instead of sleeping the current thread.

Reply:
It seemed overkill to me; A simple thread sleeping is in my opinion easier to follow than creating a rategate instance and using it in a single-threaded context.

Reply:
For now it's single threaded :) but _also_ the `RateGate` is very explicit as to what the allowables rates are and is accompanied by all the documentation and battle-tested usages. The naive thread-sleeping ignores various things like processing time and downloading time, etc, leading this process to take a _little_ bit (agreed, likely negligibly) longer.